### ISSUE TITLE
getting logged out when adding facts

### ISSUE THREAD

**Issue by LarsRabe** [fisharebest/webtrees#4717]
referring to https://www.webtrees.net/index.php/en/forum/help-for-release-2-1-x/37664-logged-out-in-problems

As was described in the forum users are experiencing regular being logged out. It happens to me on a daily base also. The browser shows a `405 `error.

I attached two parts of my web-logs where it happened. Please advice if I can debug this further.
[22_24.log](https://github.com/fisharebest/webtrees/files/10394011/22_24.log)
[21_53.log](https://github.com/fisharebest/webtrees/files/10394012/21_53.log)

**Comment by fisharebest** [comment]
```
ravenzwaaij.info anon-119-246-15-22.ip6.invalid - - [10/Jan/2023:22:22:45 +0100] "GET /tree/ravenzwaaij/add-fact/X379/OCCU HTTP/1.1" 200 17148 "https://www.ravenzwaaij.info/tree/ravenzwaaij/individual/X379/Antonij-van-Ravenswaij" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36"
ravenzwaaij.info anon-119-246-15-22.ip6.invalid - - [10/Jan/2023:22:23:32 +0100] "POST /tree/ravenzwaaij/update-fact/X379/new HTTP/1.1" 302 - "https://www.ravenzwaaij.info/tree/ravenzwaaij/add-fact/X379/OCCU" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36"
ravenzwaaij.info anon-119-246-15-22.ip6.invalid - - [10/Jan/2023:22:23:33 +0100] "GET /tree/ravenzwaaij/update-fact/X379/new HTTP/1.1" 405 20 "https://www.ravenzwaaij.info/tree/ravenzwaaij/add-fact/X379/OCCU" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36"
```

Three HTTP requests.

1) display an edit form.
2) submit the data
3) redirect to a new page after submitting form data.

The problem is that 2) can redirect to one of two places - and you are being redirected to something different.

https://github.com/fisharebest/webtrees/blob/3529c469cce04fdd2d80a72371dc283df9c2527d/app/Http/RequestHandlers/EditFactAction.php#L111-L113

Here, you are redirecting to the current URL.  This URL only accepts POST requests (i.e. form submissions).

Hence your GET request gets a 405 response.

> Please advice if I can debug this further.

Examine the source code of the edit page.  At the top of the form, there should be a hidden input, with the name `url`.  It should be a valid webtrees URL - that of the individual's page.

One possibility is that you have some "web application firewall" software running on your server that is stripping out form parameters with the name `url`, and replacing it with an empty string.  A redirect to an empty string would cause the observed behaviour.

**Comment by LarsRabe** [comment]
"web application firewall": I'll have to ask my hoster for that.

Am I right in the assumption that you added the validator on November 15th? That would explain my feeling that this problem startet with the release 2.1.8 (as some other user do also)?

**Comment by fisharebest** [comment]
Did you confirm that the hidden input contains the expected URL, and that it is sent to the server?

**Comment by fisharebest** [comment]
Validation was extended to all inputs oin 2.1.7.

**Comment by LarsRabe** [comment]
>One possibility is that you have some "web application firewall" software running on your server that is stripping out form parameters with the name `url`, and replacing it with an empty string. 
I asked my hoster, but they denied having a firewall which would do so. 


> Did you confirm that the hidden input contains the expected URL, and that it is sent to the server?
TI didn't have time to check it today. Probably this weekend.

**Comment by LarsRabe** [comment]
Again being logged out, this is was my browser showed **after** being logged out:
`https://www.ravenzwaaij.info/login/ravenzwaaij?url=https%3A%2F%2Fwww.ravenzwaaij.info%2Ftree%2Fravenzwaaij%2Fadd-spouse-to-individual%2FX430`
This is the screenshot of the event
![logout](https://user-images.githubusercontent.com/119816325/212545466-caf054f2-7e20-4b9e-9008-0858b19abc25.jpg)

**Comment by LarsRabe** [comment]
This issue is fixed by #4688
